src/cose/sign1.rs

```diff
@@ -0,0 +1,120 @@
+use crate::cose::Cose;
+use coset::cbor::Value;
+use coset::{sig_structure_data, AsCborValue, SignatureContext};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicI64, Ordering};
+use std::sync::{Arc, Mutex, OnceLock};
+
+#[derive(Clone, Debug, Default)]
+pub struct CoseSign1(pub(crate) coset::CoseSign1, Arc<Mutex<Option<i64>>>);
+
+impl CoseSign1 {
+    pub fn new(cose_sign1: coset::CoseSign1) -> Self {
+        Self(cose_sign1, Arc::new(Mutex::new(None)))
+    }
+}
+
+impl Serialize for CoseSign1 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // todo: map_err
+        self.0
+            .clone()
+            .to_cbor_value()
+            .map_err(serde::ser::Error::custom)?
+            .serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for CoseSign1 {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // Deserialize the input to a CBOR Value
+        let value = Value::deserialize(deserializer)?;
+
+        // Convert the CBOR Value to CoseSign1
+        let inner = coset::CoseSign1::from_cbor_value(value).map_err(serde::de::Error::custom)?;
+
+        Ok(CoseSign1(inner, Arc::new(Mutex::new(None))))
+    }
+}
+
+static mut SIGNATURE_PAYLOAD: OnceLock<HashMap<i64, Vec<u8>>> = OnceLock::new();
+static ATOMIC_CTR: AtomicI64 = AtomicI64::new(0);
+
+impl Cose for CoseSign1 {
+    fn signature_payload(&self) -> &[u8] {
+        // We need to return a reference to a value that is build in-place,
+        // so we need to keep it somewhere. We use a global map for this.
+        let payload = sig_structure_data(
+            SignatureContext::CoseSign1,
+            self.0.protected.clone(),
+            None,
+            &[],
+            self.0.payload.as_ref().unwrap_or(&vec![]),
```

Comment:
If the payload is `nil` (aka `None`) then the detached content needs to be provided externally, not replaced with an empty byte array.